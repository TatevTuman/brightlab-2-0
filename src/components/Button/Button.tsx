import React from 'react'
import './Button.scss'

interface ButtonProps {
  children: string
}

const Button: React.FC<ButtonProps> = props => {
  const { children } = props

  return <div className="button">{children}</div>
}

export default Button
