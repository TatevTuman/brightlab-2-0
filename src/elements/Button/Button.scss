@import "static/assets/styles/variables.scss";

button {
  height: 44px;
  padding: 12px 0 10px;
  border-radius: 22px;
  text-transform: uppercase;
  transition: all 0.2s ease;
  cursor: pointer;
  user-select: none;
  font-size: 0.75rem;

  &[data-disabled='true'] {
    background-color: var(--secondary-dark);
    border: 2px solid var(--secondary-dark);
    cursor: not-allowed;
    color: var(--white);

    &:hover, &:active {
      background-color: var(--secondary-dark);
      border: 2px solid var(--secondary-dark);
    }
  }

  @include breakpoint_max('mobile') {
    height: 32px;
    padding: 7px 0 5px;
  }
}

// Button types

@mixin primary_button($width) {
  width: 100%;
  max-width: $width;
  color: #fff;
  background-color: var(--primary);
  border: 2px solid var(--primary);

  &:hover {
    background-color: var(--secondary-dark);
    border: 2px solid var(--secondary-dark);
  }

  &:active {
    background-color: var(--black);
    border: 2px solid var(--black);
  }
}

@mixin secondary_button($width) {
  width: 100%;
  max-width: $width;
  color: var(--primary);
  background-color: transparent;
  border: 2px solid var(--primary);

  &:hover {
    background-color: var(--secondary);
    border: 2px solid var(--secondary);
  }

  &:active {
    background-color: var(--secondary);
    border: 2px solid var(--secondary);
  }
}

@mixin danger_button($width) {
  width: 100%;
  max-width: $width;
  color: var(--danger);
  background-color: transparent;
  border: 2px solid var(--danger);

  &:hover, &:active {
    background-color: var(--danger);
    color: var(--white);
  }
}

// Button sizes
$buttons_sizes: (
  "lg": 400px,
  "md": 200px,
  "sm": 100px,
);

$buttons_types_sizes: (
  "primary": $buttons_sizes,
  "secondary": $buttons_sizes,
  "danger": $buttons_sizes,
);

@each $button_type, $button_sizes in $buttons_types_sizes {
  @each $button_size_type, $button_size_value in $button_sizes {
    .button__#{$button_type}-#{$button_size_type} {

      // Button styles application

      @if $button_type == "primary" {
        @include primary_button($button_size_value);
      }

      @if $button_type == "secondary" {
        @include secondary_button($button_size_value);
      }

      @if $button_type == "danger" {
        @include danger_button($button_size_value);
      }
    }
  }
}
