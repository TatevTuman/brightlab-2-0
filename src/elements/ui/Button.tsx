import React, { forwardRef, memo } from 'react'
import cls from 'classnames'
import { Children, ClassName } from '~types'

export type ButtonVariantType = 'green' | 'green-white'

const buttonCoreVariant =
  'inline-flex items-center justify-center px-20 py-6 font-medium rounded-4 border cursor-pointer transition-colors'
const buttonGreenVariant = 'border-transparent text-white bg-green-600 hover:bg-green-700'
const buttonGreenWhiteVariant =
  'text-green-600 border-green-600 bg-white hover:text-white hover:border-green-700 hover:bg-green-700'

const getButtonVariantClass = (variant?: ButtonVariantType, className?: ClassName) =>
  cls(
    buttonCoreVariant,
    {
      [`${buttonGreenVariant}`]: !variant || variant === 'green',
      [`${buttonGreenWhiteVariant}`]: variant === 'green-white'
    },
    className
  )

export interface ButtonProps {
  className?: ClassName
  children: Children
  variant?: ButtonVariantType
  onClick?: (e: React.MouseEvent<HTMLButtonElement>) => void
}

const Button = forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => {
  const { className, variant, children, ...rest } = props

  return (
    <button ref={ref} className={getButtonVariantClass(variant, className)} {...rest}>
      {children}
    </button>
  )
})

Button.displayName = 'Button'

export default memo(Button)
