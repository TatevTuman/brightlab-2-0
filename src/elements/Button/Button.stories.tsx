import React from 'react'
import Button from './Button'

export default {
  title: 'Button',
  component: Button
}

export const Story = () => {
  return (
    <div className="story">
      <div className="story-row">
        <Button type={'primary'} size={'lg'} onClick={() => null} loading={false} disabled={false}>
          Primary Largest
        </Button>
      </div>
      <div className="story-row">
        <Button type={'secondary'} size={'md'} onClick={() => null} loading={false} disabled={false}>
          Secondary Medium
        </Button>
      </div>
      <div className="story-row">
        <Button type={'danger'} size={'sm'} onClick={() => null} loading={false} disabled={false}>
          Danger Small
        </Button>
      </div>
    </div>
  )
}
