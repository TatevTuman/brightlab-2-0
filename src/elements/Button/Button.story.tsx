import React from 'react'
import { Meta, Story } from '@storybook/react/types-6-0'
import { action } from '@storybook/addon-actions'
import Button, { ButtonProps, ButtonSizes, ButtonTypes } from './Button'
import { enumToArray, storybook } from '@utils'

interface ButtonStoryTemplateProps extends ButtonProps {}

export default {
  title: 'Elements/Button',
  argTypes: {
    children: { control: 'text', name: 'Text' },
    type: { control: storybook.controls.select(enumToArray(ButtonTypes)), name: 'Type' },
    size: { control: storybook.controls.select(enumToArray(ButtonSizes)), name: 'Size' },
    to: { control: 'text', name: 'Link' },
    disabled: { control: 'boolean', name: 'Disabled' },
    loading: { control: 'boolean', name: 'Loading' },
    centered: { control: 'boolean', name: 'Centered' },
    onClick: { action: 'onClick' },
    form: storybook.args.disabled,
    submit: storybook.args.disabled,
    className: storybook.args.disabled
  }
} as Meta

const ButtonStoryTemplate: Story<ButtonStoryTemplateProps> = args => {
  return <Button {...args} />
}

ButtonStoryTemplate.args = {
  children: 'Text',
  type: ButtonTypes.primary,
  size: ButtonSizes.lg,
  onClick: action('onClick')
}

export const ButtonStory = ButtonStoryTemplate.bind({})
ButtonStory.args = {
  ...ButtonStoryTemplate.args
}
