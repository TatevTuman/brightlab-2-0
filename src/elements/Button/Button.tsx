import React, { memo, MouseEvent } from 'react'
import { navigate } from 'gatsby'
import { useDelayEffect } from '@hooks'
import './Button.scss'

type ButtonKeyboardEvent = React.KeyboardEvent<HTMLButtonElement>
type ButtonMouseEvent = MouseEvent<HTMLButtonElement>
type ButtonEvents = ButtonKeyboardEvent | ButtonMouseEvent

export interface ButtonProps {
  children: string
  className?: string
  type?: string
  size?: string
  onClick?(): void
  disabled?: boolean
  loading?: boolean
  to?: string
  submit?: boolean
  centered?: boolean
}

const Button: React.FC<ButtonProps> = props => {
  const { className, type, size, to, children, onClick, loading, disabled, submit, centered } = props

  const classNames = `button button__${type}-${size} ${className}`
  const delayedLoading = useDelayEffect(loading!)

  const handleClick = async (e: ButtonEvents) => {
    e.stopPropagation()

    if (delayedLoading || disabled) return

    if (to) await navigate(to)
    else onClick!()
  }

  const handleKeyDown = async (e: ButtonKeyboardEvent) => {
    const isEnterKey = e.key === 'Enter'

    if (isEnterKey) {
      await handleClick(e)
    }
  }

  return (
    <button
      className={classNames}
      type={submit ? 'submit' : 'button'}
      data-disabled={delayedLoading || disabled}
      data-centered={centered}
      onClick={handleClick}
      onKeyDown={handleKeyDown}
      tabIndex={0}
    >
      {delayedLoading ? 'Загрузка...' : children}
    </button>
  )
}

Button.defaultProps = {
  className: '',
  type: 'primary',
  size: 'lg',
  onClick: () => null,
  loading: false,
  disabled: false,
  submit: false,
  centered: false
}

export default memo(Button)
