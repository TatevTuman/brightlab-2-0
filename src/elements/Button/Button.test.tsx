import React from 'react'
import { render, cleanup, waitFor, fireEvent, act } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Button from './Button'

jest.useFakeTimers()

// beforeAll(() => {})
// afterAll(() => {})
// beforeEach(() => {})
afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

describe('Button', () => {
  const props = {
    children: 'children',
    onClick: jest.fn()
  }

  it('renders correctly', async () => {
    const { container, getByText } = render(<Button {...props} />)
    await waitFor(() => container)

    expect(container).toMatchSnapshot()
    expect(getByText(props.children)).toBeInTheDocument()
  })

  it('renders loading correctly', async () => {
    const { container, rerender } = render(<Button {...props} loading={true} />)
    await waitFor(() => container)

    const button = container.querySelector('button')
    let loader = container.querySelector('.loader')

    expect(loader).toBeInTheDocument()
    expect(button).toHaveAttribute('data-disabled', 'true')

    rerender(<Button {...props} loading={false} />)

    setTimeout(() => {
      loader = container.querySelector('.loader')

      expect(loader).toBeNull()
      expect(button).toHaveAttribute('data-disabled', 'false')
    }, 1000)

    act(() => {
      jest.runAllTimers()
    })
  })

  it('renders disabled correctly', () => {
    const { getByText, rerender } = render(<Button {...props} disabled={true} />)
    const button = getByText(props.children)

    expect(button).toHaveAttribute('data-disabled', 'true')

    rerender(<Button {...props} disabled={false} />)

    expect(button).toHaveAttribute('data-disabled', 'false')
  })

  it('click', () => {
    const { container } = render(<Button {...props} />)
    const button = container.querySelector('button')

    userEvent.click(button!)

    expect(props.onClick).toHaveBeenCalledTimes(1)
  })

  it('click on loading', () => {
    const { container } = render(<Button {...props} loading={true} />)
    const button = container.querySelector('button')

    userEvent.click(button!)

    expect(props.onClick).toHaveBeenCalledTimes(0)
  })

  it('click on disabled', () => {
    const { container } = render(<Button {...props} disabled={true} />)
    const button = container.querySelector('button')

    userEvent.click(button!)

    expect(props.onClick).toHaveBeenCalledTimes(0)
  })

  it('navigate instead of click', () => {
    const { container } = render(<Button {...props} to={'/'} />)
    const button = container.querySelector('button')

    userEvent.click(button!)

    expect(props.onClick).toHaveBeenCalledTimes(0)
  })

  it('onkeydown', () => {
    const { container } = render(<Button {...props} />)
    const button = container.querySelector('button')

    userEvent.tab()

    fireEvent.keyDown(button!, { key: 'Tab' })
    expect(props.onClick).toHaveBeenCalledTimes(0)

    fireEvent.keyDown(button!, { key: 'Esc' })
    expect(props.onClick).toHaveBeenCalledTimes(0)

    fireEvent.keyDown(button!, { key: 'Enter' })
    expect(props.onClick).toHaveBeenCalledTimes(1)
  })
})
