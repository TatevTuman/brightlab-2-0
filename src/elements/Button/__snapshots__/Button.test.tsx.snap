// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<div>
  <button
    class="button button__primary-lg "
    data-centered="false"
    data-disabled="false"
    tabindex="0"
    type="button"
  >
    children
  </button>
</div>
`;
